src/main/scala/org/kframework/minikore/interfaces/MiniKoreInterface.scala

```diff
@@ -0,0 +1,303 @@
+package org.kframework.minikore.interfaces
+
+object TreeInterface {
+
+  import PatternInterface._
+
+  sealed trait AST
+
+
+  sealed trait Node extends AST with Product {
+    def args: Seq[Pattern]
+
+    def build(children: Seq[Pattern]): Pattern
+  }
+
+  object Node {
+    def unapply(arg: Node): Option[Seq[Pattern]] = Some(arg.args)
+  }
+
+
+  sealed trait Leaf[C] extends AST with Product {
+    def contents: C
+
+    def build(contents: C): Pattern
+  }
+
+  object Leaf {
+    def unapply[C](arg: Leaf[C]): Option[C] = Some(arg.contents)
+  }
+
+
+  sealed trait Leaf2[CC1, CC2] extends Leaf[Product2[CC1, CC2]] with Product2[CC1, CC2] {
+    override def contents: Product2[CC1, CC2] = (_1, _2)
+
+    def apply(_1: CC1, _2: CC2): Pattern
```

Comment:
I agree with Cosmin that naming it `apply` is confusing (especially for the example that he gave with `Rewrite`), but I think that renaming it to `build` is a fine solution. Though for the implementations it would be nice if the `apply` method could be used to actually build them (so it's easier to construct them), which means we'll just be adding the `apply` back in at that level.

I'm still opposed to the builder hierarchy because we were creating a whole hierarchy just to wrap the types `Seq[Pattern] => _`, which is already a covariant functor supported directly by the `Function1` trait. So there's no need to re-create some specific wrapper of `Function1` for our needs when it already exists.

Also, when looking at the diff, all you see is red, so we know that this design is more compact. I would also argue it's much easier to understand, but that's subjective.